### ISSUE TITLE
KeyError or AttributeError when I do a `doc.register()`

### ISSUE THREAD

**Issue by AbdealiJK** [jmcarp/flask-apispec#116]
I am trying my own example after reading the docs:

```
import os

from flask import Flask, Blueprint
from flask_apispec import FlaskApiSpec, MethodResource, marshal_with
from flask_marshmallow import Marshmallow, Schema
from marshmallow import fields


APP_DIR = os.path.dirname(os.path.abspath(__file__))


class BookSchema(Schema):
    id = fields.Int(dump_only=True)
    name = fields.Str(dump_only=True)


@marshal_with(BookSchema(many=True))
class BookListResource(MethodResource):

    def get(self):
        return [{"id": 1, "name": 30}]


def create_app():
    # Initialize flask application
    flask_app = Flask(__name__, instance_relative_config=True)
    ma = Marshmallow(flask_app)
    docs = FlaskApiSpec(flask_app)

    book_bp = Blueprint('books_api', __name__, url_prefix='/books/')
    book_bp.add_url_rule('/', view_func=BookListResource.as_view('BookList'))

    app = flask_app
    app.register_blueprint(book_bp)
    docs.register(BookListResource, blueprint="books_api")
    return flask_app


if __name__ == '__main__':
    app = create_app()
    app.run(host='0.0.0.0', debug=True, port=9000)

```

I get the following error:

```
Traceback (most recent call last):
  File "a.py", line 39, in <module>
    app = create_app()
  File "a.py", line 34, in create_app
    docs.register(BookListResource, blueprint="books_api")
  File "/Users/abdealijk/Documents/projects/flaskapi/venv/lib/python3.6/site-packages/flask_apispec/extension.py", line 119, in register
    resource_class_args, resource_class_kwargs)
  File "/Users/abdealijk/Documents/projects/flaskapi/venv/lib/python3.6/site-packages/flask_apispec/extension.py", line 66, in _defer
    bound()
  File "/Users/abdealijk/Documents/projects/flaskapi/venv/lib/python3.6/site-packages/flask_apispec/extension.py", line 141, in _register
    resource_class_kwargs=resource_class_kwargs,
  File "/Users/abdealijk/Documents/projects/flaskapi/venv/lib/python3.6/site-packages/flask_apispec/apidoc.py", line 37, in convert
    rules = self.app.url_map._rules_by_endpoint[endpoint]
KeyError: 'books_api.booklistresource'
```

On changing the line:
`book_bp.add_url_rule('/', view_func=BookListResource.as_view('BookList'))`
to:
`book_bp.add_url_rule('/', view_func=BookListResource.as_view('booklistresource'))`

it gives another error:
```
['books_api.booklistresource']
Traceback (most recent call last):
  File "a.py", line 40, in <module>
    app = create_app()
  File "a.py", line 35, in create_app
    docs.register(BookListResource, blueprint="books_api")
  File "/Users/abdealijk/Documents/projects/flaskapi/venv/lib/python3.6/site-packages/flask_apispec/extension.py", line 119, in register
    resource_class_args, resource_class_kwargs)
  File "/Users/abdealijk/Documents/projects/flaskapi/venv/lib/python3.6/site-packages/flask_apispec/extension.py", line 66, in _defer
    bound()
  File "/Users/abdealijk/Documents/projects/flaskapi/venv/lib/python3.6/site-packages/flask_apispec/extension.py", line 146, in _register
    self.spec.add_path(**path)
  File "/Users/abdealijk/Documents/projects/flaskapi/venv/lib/python3.6/site-packages/apispec/core.py", line 269, in add_path
    plugin.operation_helper(path=path, operations=path.operations, **kwargs)
  File "/Users/abdealijk/Documents/projects/flaskapi/venv/lib/python3.6/site-packages/apispec/ext/marshmallow/__init__.py", line 250, in operation_helper
    self.resolve_schema(response)
  File "/Users/abdealijk/Documents/projects/flaskapi/venv/lib/python3.6/site-packages/apispec/ext/marshmallow/__init__.py", line 121, in resolve_schema
    data['schema'] = self.openapi.resolve_schema_dict(data['schema'])
  File "/Users/abdealijk/Documents/projects/flaskapi/venv/lib/python3.6/site-packages/apispec/ext/marshmallow/openapi.py", line 696, in resolve_schema_dict
    return self.schema2jsonschema(schema, dump=dump)
  File "/Users/abdealijk/Documents/projects/flaskapi/venv/lib/python3.6/site-packages/apispec/ext/marshmallow/openapi.py", line 561, in schema2jsonschema
    fields, schema, use_refs=use_refs, dump=dump, name=name,
  File "/Users/abdealijk/Documents/projects/flaskapi/venv/lib/python3.6/site-packages/apispec/ext/marshmallow/openapi.py", line 627, in fields2jsonschema
    observed_field_name = self._observed_name(field_obj, field_name)
  File "/Users/abdealijk/Documents/projects/flaskapi/venv/lib/python3.6/site-packages/apispec/ext/marshmallow/openapi.py", line 129, in _observed_name
    return field.data_key or name
AttributeError: 'Integer' object has no attribute 'data_key'
```

Here's my stack:
```
Flask                      0.12
flask-apispec              0.7.0
flask-marshmallow          0.8.0
marshmallow                2.15.0
```

**Comment by AbdealiJK** [comment]
After reading through the code  a bit I found an `endpoint` argument which allows me to do:
```
docs.register(BookListResource, blueprint="books_api", endpoint='BookList')
```
So, that works now.

For the second error, I found that the error was happening in `flaskapi.ext.marshmallow`
And that `data_key` is present in marshmallow 3.x [link](https://marshmallow.readthedocs.io/en/3.0/upgrading.html#load-from-and-dump-to-are-merged-into-data-key)

Found: https://github.com/marshmallow-code/apispec/issues/200 in the apispec repo.

Seems like my pip was reporting 2.x but it was actually 3.x when i do `python -c "import marshmallow; print(marshmallow.__version__)"`
Eitherway, seems like a `apispec` issue.
